### ISSUE TITLE
Support for `ser2net` connection to smart meter

### ISSUE THREAD

**Issue by hemant5400z** [ebaauw/homebridge-p1#55]
In have a p1 exposes using ser2net http which is connected to domoticz. I want to use the same connection to expose it to homekit so insteas of serial use http. Would that be possible. As my p1 smartmeter is on a remote system not close to domoticz nor to homebridge hencd tty is not an option.

Cheers

**Comment by ebaauw** [comment]
I don’t understand your setup.  Afaik, `ser2net` doesn’t talk `http`.  Also using `http` for the P1 telegrams makes no sense to me, as the smart meter pushes these.  Homebridge P1 doesn’t request (get) telegrams; instead it expects to receive notifications.

**Comment by hemant5400z** [comment]
Hi,

Setup is as follwos i have a mini pc which serves as p1 server. The p1 is connected to its ttyusb port. Ser2net converts this to ip:port. I use this to connect to my domoticz which has a p1 over land connection type. I want to do similar which homebridge instead of serial go over the lan connection

![image](https://user-images.githubusercontent.com/16900394/132704076-838af286-996a-4296-ad93-b35ff557cb12.jpeg)

Like this.

**Comment by ebaauw** [comment]
What is your `ser2net` config?

**Comment by ebaauw** [comment]
I have Homebridge P1 beta v1.2.10-0 working with `ser2net`.  Could you please try that version?  Set `serialPort` to _hostname_`:`_port_ of the server running `ser2net`.  In `ser2net.conf`, you need to setup a raw connection (in my case connecting port 2000 to /dev/ttyUSB0):
```
2000:raw:600:/dev/ttyUSB0:115200 8DATABITS NONE 1STOPBIT
```
I find the connection considerably less reliable than using a direct connection.  Not sure if that's `ser2net` or my WiFi.  I had to add checksum validation, discarding telegrams with a CRC error.  I never had a need for that using the serial port directly.

Note that DSMR 2.2 probably won't work with `ser2net`.  Not only does it use 9600 7E1 instead of 115200 8N1, it also doesn't provide checksums, so I cannot check wether a telegram is valid.  You'll get all sorts of errors when Homebridge P1 tries to parse an invalid telegram.

**Comment by hemant5400z** [comment]
Hi ebauuw,

Let me test that i have the dsmr 5 version. Also in domoticz i use crc running this setup for 8 months now without a problem. I gues need to uninstall current version? How do i pull the beta as it is not showing in the search box

**Comment by ebaauw** [comment]
> Let me test that i have the dsmr 5 version

The version is usually mentioned on a label on the meter.  Also if you receive a telegram every second, it's DSMR 5 alright.  Otherwise, Homebridge P1 logs the version on startup (after the meter's serial number).

> How do i pull the beta as it is not showing in the search box

Use the _Install Previous Version_ option from the UI, under the wrench symbol in the _Plugins_ screen.

> I find the connection considerably less reliable than using a direct connection. Not sure if that's ser2net or my WiFi.

Left two instances running overnight: one on the Pi with the USB P1 cable (`pi4`), using `/dev/ttyUSB0` and one on a remote Pi, using `pi4:2000`.  The local Pi showed shy of 300 CRC errors, the remote Pi almost nine times as many.  I had configured `ser2net` months ago already and haven't seen any errors due to parsing invalid telegrams.  I would theorise the issues are caused by multiple concurrent connections to the serial port.

I now reverted to connecting the remote Pi over Homebridge P1's websocket relay to see if the local Pi will still be reporting CRC errors.  Next step will be connecting the local Pi over `ser2net` (using the loopback interface).  And after that, connecting both local and remote Pi over `ser2net` (hoping it would share a single connection to the serial port between multiple TCP clients).

**Comment by hemant5400z** [comment]
Hi just installed.

I can see the messages but seems strange device in homekit.

No actual numbers but a programmable switch?

![image](https://user-images.githubusercontent.com/16900394/132820029-0bf24380-4877-443b-865e-21f18ddb4fd7.jpeg)

**Comment by hemant5400z** [comment]
![image](https://user-images.githubusercontent.com/16900394/132820090-aa6f9c46-a3c4-4b40-97aa-86158748382c.jpeg)
![image](https://user-images.githubusercontent.com/16900394/132820124-b1dd8f39-cffc-40e0-a0e9-f5c9b8ada5f7.jpeg)

**Comment by ebaauw** [comment]
Use a decent HomeKit app.

**Comment by ebaauw** [comment]
Some further analysis: also seeing "invalid telegram" and "no data received" warnings in the log of the local Pi and remote Pi.  Total warnings (incl. CRC errors) about 1100 on the local Pi, and 3100 on the remote Pi.  The warnings on the two Pi are independent (each individual warning only shows on one of the Pi-s).  I haven't seen any warnings on for over half an hour, after reverting to the serial connection.  Will be trying a local `ser2net` connection next.

**Comment by hemant5400z** [comment]
Ok installed eve and indeed working as expected. In my mind i thought it would be exposed in homekit.

I connot detect errors in the log not sure why you get them. Min is going over cable connection. I have 4 simultanous connections and all work fine.

**Comment by hemant5400z** [comment]
Also i'm not running on a pi. I have ubuntu lts 20.04 running on a virtual machine. The usb is passed on to the vm with passthrough.

**Comment by hemant5400z** [comment]
Just to be more specific:

Server Ubuntu LTS 20.04 running on ESXI.
P1: connection to Meter about 10 meters from it using USB2serial (home made :) )
Ser2net captures ttyUSB0 to ip:port max 10 connections
Connection 1 Docker : DSMR
Connection 2 Docker : Domoticz
Connection 3 Azure cloud VM : DSMR cloud instance
Connection  4 Docker : Homebridge

Why? for fun :)

**Comment by ebaauw** [comment]
>  In my mind i thought it would be exposed in homekit.

It is.  Apple's Home app does not equal HomeKit.

> Ser2net captures ttyUSB0 to ip:port max 10 connections

Where/how do you specify the max connections?  I tried to open the `ser2net` port from two clients, but it closes the second connection on me.

**Comment by ebaauw** [comment]
Google is my friend.  Changed `ser2net.conf` to read:
```
2000:raw:600:/dev/ttyUSB0:115200 8DATABITS NONE 1STOPBIT max-connections=3
```
Now the second client stays connected, and both clients receive data.

**Comment by hemant5400z** [comment]
Mine is slight different also max-num only works with v3.x not supported in V2.

ipv4,tcp,5000:raw:0:/dev/ttyUSB-P1:115200 NONE 1STOPBIT 8DATABITS max-connections=10

you are using 600 which drops connection in 10min i think, domoticz did not like that.

**Comment by ebaauw** [comment]
> you are using 600 which drops connection in 10min i think, domoticz did not like that.

Only when it doesn't see any telegrams for 10 minutes, in my experience.  This would only happen when the cable has been disconnected from the smart meter.  Homebridge P1 re-connects automatically some time after a connection has been closed.

**Comment by ebaauw** [comment]
Now going 40 minutes without warnings, using two concurrent `ser2net` connections.  I'm concluding that the instability was caused by two concurrent connections (Homebridge P1 and `ser2net`) to the serial port.

**Comment by ebaauw** [comment]
Has been running stable for over a week.  Published v1.2.10.

**Comment by hemant5400z** [comment]
That is very cool!

Verstuurd vanaf mijn iPhone

Op 17 sep. 2021 om 10:33 heeft Erik Baauw ***@***.***> het volgende geschreven:

﻿

Has been running stable for over a week. Published v1.2.10.
